### DISCUSSION TITLE
response 404 from aws s3 for filenames with spaces

### DISCUSSION THREAD

**Discussion by cpeled** [nginxinc/nginx-s3-gateway#79] [Q&A] [answered]
Hi.
in AWS, when filename with spaces is uploaded, in the generated URL, every space replaced with '+'.
example: i uploaded a filename: "n ginx.conf".
the URL generated by aws: https://<BUCKET>.s3.eu-central-1.amazonaws.com/n+ginx.conf

when i try to download the file using s3 gw using path "n+ginx.conf", i get 404.

since the  GW replaces '+' with '%2B', i made a try and in s3gateway.ls, in "_encodeURIComponent", i replaced back from '%2B' to '+', but then i got from AWS "SignatureDoesNotMatch" response, and in the details i see that the '+' was changed to '%20' which is the space code.

can you please help me understand the issue?

Best,
Chen

**Comment by dekobon** [comment]
Hi Chen,

I transferred this from an issue to a discussion because I do not believe it is an issue. The gateway is doing the correct thing by escaping plus signs as `%2B`. I think there is a problem in the service that is uploading files to S3 because spaces should be encoded as `%20` and not `+`.
See [this StackOverflow](https://stackoverflow.com/questions/2678551/when-should-space-be-encoded-to-plus-or-20) conversation for some more detail.

In the automated tests for the gateway, we [explicitly test filenames containing pluses](https://github.com/nginxinc/nginx-s3-gateway/blob/master/test/integration/test_api.sh#L260). As noted by the comment, a plus must be encoded as `%2B`.

I hope this makes sense. I'm afraid that my explanation might be a bit confusing. Anyways, I think the problem is that your files that are uploaded to S3 should have `%20` for spaces and not the `+` sign.

**Reply by cpeled** [reply]
hi @dekobon and thanks for your response.

i tried to upload the file from:
1. aws s3 console
2. aws s3 cli.
3. aws boto3 python client

all resulted the same - the generated url **set by aws** was with '+' instead of space.

do you suggest to encode the filename before uploading? i upload it as is, means with ' ' and not '+'.

Chen

**Comment by dekobon** [comment, marked as answer]
Hi Chen,

I did the following verifications:
```
# Upload file with + in it to s3
$ echo "Hello World" > "file_with_+.txt"
$ aws s3 cp "file_with_+.txt" s3://testbucket/
upload: ./file_with_+.txt to s3://testbucket/file_with_+.txt

# Upload file with space in it to s3
$ aws s3 cp "file_with space.txt" s3://testbucket/
upload: ./file_with space.txt to s3://testbucket/file_with space.txt

# Run the container in another terminal
$ docker run --env-file ./test-settings --rm -it --publish 80:80 --name nginx-s3-gateway nginx-s3-gateway

# Download uploaded file with + in it escaped properly
$ curl "http://localhost/file_with_%2B.txt"
Hello World

# Verify that we can download uploaded file with + in it not escaped properly
$ curl "http://localhost/file_with_+.txt"
Hello World

# Now, let's try it again to make absolutely sure that curl is not escaping the request
$ nc localhost 80
GET /file_with_+.txt HTTP/1.1
Host: localhost
User-Agent: curl/7.68.0
Accept: */*

HTTP/1.1 200 OK
Server: nginx
Date: Wed, 14 Dec 2022 18:50:52 GMT
Content-Type: text/plain
Content-Length: 12
Connection: keep-alive
Last-Modified: Wed, 14 Dec 2022 18:44:36 GMT
ETag: "e59ff97941044f85df5297e1c302d260"
Accept-Ranges: bytes

Hello World

# Download uploaded file with space in it escaped properly
$ curl "http://localhost/file_with%20space.txt"
Hello World

# Verify that we cannot download uploaded file with space in it not escaped properly
$ curl "http://localhost/file_with space.txt"
<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<hr><center>nginx</center>
</body>
</html>

# Let's try plus and see what happens
$ curl "http://localhost/file_with+space.txt"
<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<hr><center>nginx</center>
</body>
</html>

# It appears to have the same behavior - as both are invalid URLs
# Let's see what the s3 client does:
$ aws s3 --debug cp "s3://testbucket/file_with space.txt" /tmp/
# It outputs a lot of debug information, but the important part is:
# 2022-12-14 11:00:17,724 - ThreadPoolExecutor-0_0 - urllib3.connectionpool - DEBUG - https://testbucket.s3.amazonaws.com:443 "GET /file_with%20space.txt HTTP/1.1" 200 12
# That shows that AWS is using %20 to escape spaces
```

Does this clarify the behavior for you?

**Reply by cpeled** [reply]
Hi @dekobon ,
thank you so much for this clarification!

it helped me to find the issue in my system that was indeed in the caller side and not the gw side.


Best,
Chen
